frontend/renew_proxies.py

```diff
@@ -0,0 +1,151 @@
+           '-cert', proxy.cert,
+           '-key', proxy.key,
+           '-out', proxy.output + '.new',
+           '-hours', proxy.lifetime,
+           '-voms', vo_info.name,
+           '-hostcert', vo_info.cert,
+           '-hostkey', vo_info.key,
+           '-uri', voms_uri,
+           '-fqan', vo_info.fqan]
+    return _run_command(cmd)
+
+def main():
+    """Main entrypoint
+    """
+    config = ConfigParser.ConfigParser(DEFAULTS)
+    config.read(CONFIG)
+    proxies = config.sections()
+
+    # Verify config sections
+    if proxies.count('FRONTEND') != 1:
+        raise RuntimeError("ERROR: there must be only one [FRONTEND] section in %s" % CONFIG)
+    if len([x for x in proxies if x.startswith('PILOT')]) < 1:
+        raise RuntimeError("ERROR: there must be at least one [PILOT] section in %s" % CONFIG)
+
+    # Proxies need to be owned by the 'frontend' user
+    try:
+        fe_user = pwd.getpwnam('frontend')
+    except KeyError:
+        raise RuntimeError("ERROR: missing 'frontend' user")
+
+    # Load VOMS Admin server info for case-sensitive VO name and for faking the VOMS Admin server URI
+    with open('/etc/vomses', 'r') as _:
+        vo_info = re.findall(r'"(\w+)"\s+"([^"]+)"\s+"(\d+)"\s+"([^"]+)"', _.read(), re.IGNORECASE)
+        vomses = dict([(vo[0].lower(), {'name': vo[0], 'uri': vo[1] + ':' + vo[2], 'subject': vo[3]}) for vo in vo_info])
+
+    # Proxy renewals
+    for proxy_section in proxies:
+        proxy_config = dict(config.items(proxy_section))
+        proxy = Proxy(proxy_config['proxy_cert'], proxy_config['proxy_key'],
+                      proxy_config['output'], proxy_config['lifetime'])
+
+        if int(proxy.lifetime)*3600 - proxy.timeleft() < int(proxy_config['frequency'])*3600:
+            print 'Skipping renewal of %s: time remaining within the specified frequency' % proxy.output
+            continue
+
+        if proxy_section == 'FRONTEND':
+            stdout, stderr, retcode = voms_proxy_init(proxy)
+        elif proxy_section.startswith('PILOT'):
+            voms_info = vomses[proxy_config['vo'].lower()]
+            vo_attr = VO(voms_info['name'], proxy_config['fqan'])
+
+            if proxy_config['use_voms_admin'].lower() == 'true':
+                # we specify '-order' because some European CEs care about VOMS AC order
+                stdout, stderr, retcode = voms_proxy_init(proxy, '-voms', vo_attr.voms, '-order', vo_attr.fqan)
+            else:
+                vo_attr.cert = proxy_config['vo_cert']
+                vo_attr.key = proxy_config['vo_key']
+                stdout, stderr, retcode = voms_proxy_fake(proxy, vo_attr, voms_info['uri'])
+        else:
+            print "WARNING: Unrecognized configuration section %s found in %s.\n" % (proxy, CONFIG) + \
+                "Valid configuration sections: 'FRONTEND' or 'PILOT'."
+
+        if retcode == 0:
+            proxy.write()
+            os.chown(proxy.output, fe_user.pw_uid, fe_user.pw_gid)
```

Comment:
This has a race condition that may end in tears.  Do this instead:

1.  Create the proxy in a temporary filename (preferably using a class from the `tempfile` library).  Should be in the same directory as the output in order to avoid cross-filesystem renames later (keeps the rename atomic).
2.  Do the appropriate `chown`
3.  Write out the proxy, fsync the file descriptor, close the fd.
4.  Use `os.rename` to overwrite the new proxy with the old one.

This way, you're guaranteed that the file at `proxy.output` always is valid.

Reply:
Isn't 3 already handled by 1? I can give the various `voms-proxy-*` commands the tempfile as output but I don't get choose when the fd writing/handling occurs.